File: main/sherlock-and-valid-string/sherlock-and-valid-string.rb
def valid? s
	# letter in s -> how many times it appears in s
	charfreqs = Hash.new 0
	0.upto(s.size - 1) {|i| charfreqs[s[i]] += 1}

	# number of times -> number of distinct letters appearing that often
	freqfreqs = Hash.new 0
	charfreqs.values.each {|freq| freqfreqs[freq] += 1}

	# 0 or 1 frequencies: valid; more than 2 frequencies: invalid
	return freqfreqs.count < 2 unless freqfreqs.count == 2

	# if we just have to remove 1 occurrence of 1 letter: valid
	return true if freqfreqs[1] == 1

	# if one letter appears one time more often than each other letter,
	# then a single occurrence of that more frequent letter may be removed
	# so all letters appear equally frequently -- such a string is valid
	lowfreq, highfreq = freqfreqs.keys.min, freqfreqs.keys.max
	return highfreq - lowfreq == 1 && freqfreqs[highfreq] == 1
end

puts valid?(gets.chomp) ? 'YES' : 'NO'

File: main/sherlock-and-valid-string/sherlock-and-valid-string.cs
#define DEBUG

using System;
using System.Collections.Generic;
using System.Linq;

internal static class Solution {
    private static void Increment<TKey>(this Dictionary<TKey, int> self,
                                        TKey key)
    {
        if (self.ContainsKey(key))
            ++self[key];
        else
            self.Add(key, 1);
    }

    private static void Exchange<T>(T a, T b)
    {
        var c = a;
        a = b;
        b = c;
    }

    private static bool IsDeemedValid(this string s)
    {
        var freqs = new Dictionary<char, int>();
        foreach (var ch in s) freqs.Increment(ch);

        switch (freqs.Values.Distinct().Count()) {
        case 0: case 1:
            return true;

        case 2:
            // majority: the number of times each of most chars appear in s
            // minority: how many times each of the other chars appear in s

            var freq = freqs.Values.First();
            var majority = freqs.Values.Where(f => f == freq).Count();
            var minority = freqs.Count - majority;
            if (majority < minority) Exchange(majority, minority);



        default:
            return false;
        }
    }

    private static void Main()
    {
        Console.WriteLine(Console.ReadLine().IsDeemedValid() ? "YES" : "NO");
    }
}
